File: android/src/main/kotlin/my/com/fromyourlover/zendesk_plus/ZendeskPlusPlugin.kt
package my.com.fromyourlover.zendesk_plus

import android.app.Activity
import android.content.Context
import android.graphics.Color
import io.flutter.Log

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import my.com.fromyourlover.pigeon.FlutterError
import my.com.fromyourlover.pigeon.ZendeskHostApi
import my.com.fromyourlover.pigeon.ZendeskListener
import my.com.fromyourlover.pigeon.ZendeskUser
import zendesk.android.Zendesk
import zendesk.android.ZendeskResult
import zendesk.android.events.ZendeskEvent
import zendesk.android.events.ZendeskEventListener
import zendesk.android.events.exception.ZendeskJwtExpiredException
import zendesk.android.messaging.model.UserColors
import zendesk.messaging.android.DefaultMessagingFactory
import zendesk.logger.Logger;

/** ZendeskPlusPlugin */
class ZendeskPlusPlugin : FlutterPlugin, ActivityAware, ZendeskHostApi {

    private var activity: Activity? = null // Store the activity
    private var context: Context? = null
    private var zendesk: Zendesk? = null

    private var flutterApi: ZendeskListener? = null

    //    Customization
    var lightColors: UserColors? = null
    var darkColors: UserColors? = null

    private val zendeskEventListener: ZendeskEventListener = ZendeskEventListener { zendeskEvent ->


        when (zendeskEvent) {
            is ZendeskEvent.AuthenticationFailed -> {
                flutterApi?.onEvent(my.com.fromyourlover.pigeon.ZendeskEvent.AUTHENTICATION_FAILED) { }

                when (zendeskEvent.error) {
                    is ZendeskJwtExpiredException -> flutterApi?.onEvent(my.com.fromyourlover.pigeon.ZendeskEvent.JWT_EXPIRED_EXCEPTION) {}
                }
            }

            is ZendeskEvent.ConnectionStatusChanged -> flutterApi?.onEvent(my.com.fromyourlover.pigeon.ZendeskEvent.CONNECTION_STATUS_CHANGED) { }
            is ZendeskEvent.ConversationAdded -> flutterApi?.onEvent(my.com.fromyourlover.pigeon.ZendeskEvent.CONNECTION_STATUS_CHANGED) { }
            is ZendeskEvent.FieldValidationFailed -> flutterApi?.onEvent(my.com.fromyourlover.pigeon.ZendeskEvent.FIELD_VALIDATION_FAILED) { }
            is ZendeskEvent.SendMessageFailed -> flutterApi?.onEvent(my.com.fromyourlover.pigeon.ZendeskEvent.SEND_MESSAGE_FAILED) { }
            is ZendeskEvent.UnreadMessageCountChanged -> flutterApi?.onEvent(my.com.fromyourlover.pigeon.ZendeskEvent.UNREAD_MESSAGE_COUNT_CHANGED) { }
        }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        Log.i("Zendesk", "Setup Method Chanel")
        ZendeskHostApi.setUp(binding.binaryMessenger, this)
        flutterApi = ZendeskListener(binding.binaryMessenger)

        context = binding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        ZendeskHostApi.setUp(binding.binaryMessenger, null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity // Get the activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null // Clear the activity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity // Re-get the activity
    }

    override fun onDetachedFromActivity() {
        activity = null // Clear the activity
    }

    override fun initialize(
        androidChannelId: String?, iosChannelId: String?, callback: (Result<Unit>) -> Unit
    ) {
        if (androidChannelId == null) {
            callback(
                Result.failure(
                    FlutterError(
                        "missing-argument",
                        "Android app ID is required",
                        "No androidAppId provided for initialization"
                    )
                )
            )
            return
        }

        if (context == null) {
            callback(
                Result.failure(
                    FlutterError(
                        "invalid-context",
                        "Android context is not available",
                        "Make sure plugin is properly initialized with context"
                    )
                )
            )
            return
        }

        CoroutineScope(Dispatchers.Main).launch {
            try {
                Log.i("Zendesk", "Starting Zendesk initialization...")

                val result = Zendesk.initialize(
                    context = context!!,
                    channelKey = androidChannelId,
                    messagingFactory = DefaultMessagingFactory(lightColors, darkColors)
                )

                when (result) {
                    is ZendeskResult.Success -> {
                        Log.i("Zendesk", "Initialization successful")
                        zendesk = result.value

                        callback(Result.success(Unit))
                    }

                    is ZendeskResult.Failure -> {
                        Log.e("Zendesk", "Initialization failed: ${result.error}")
                        callback(
                            Result.failure(
                                FlutterError(
                                    "zendesk-error",
                                    "Zendesk initialization failed",
                                    result.error.message ?: "Unknown error"
                                )
                            )
                        )
                    }
                }
            } catch (e: Exception) {
                Log.e("Zendesk", "Exception during initialization", e)
                callback(
                    Result.failure(
                        FlutterError(
                            "unexpected-error",
                            "Unexpected error occurred",
                            e.message ?: "No error details"
                        )
                    )
                )
            }
        }
    }


    override fun openChat() {

        checkInitialization()

        CoroutineScope(Dispatchers.Main).launch {
            try {
                if (activity != null) {
                    zendesk!!.messaging.showMessaging(activity!!)
                } else {
                    println("Activity is null. Cannot show Zendesk messaging.")
                    // Handle the error appropriately, e.g., send an error back to Flutter
                }
            } catch (e: Exception) {
                println("Error showing Zendesk messaging: ${e.message}")
                // Handle errors
            }
        }
    }

    override fun signIn(jwt: String, callback: (Result<ZendeskUser>) -> Unit) {

        checkInitialization()

        CoroutineScope(Dispatchers.Main).launch {
            try {
                // Call the suspend function and await its result
                // Handle ZendeskResult and map it to Result<String>
                when (val result = zendesk!!.loginUser(jwt)) {
                    is ZendeskResult.Success -> {
                        // Extract data from ZendeskUser (replace with actual logic)
                        val userData = result.value
                        callback(
                            Result.success(
                                ZendeskUser(
                                    id = userData.id, externalId = userData.externalId!!
                                )
                            )
                        )
                    }

                    is ZendeskResult.Failure -> {
                        callback(Result.failure(result.error))
                    }
                }
            } catch (e: Exception) {
                // Handle exceptions (e.g., network errors)
                println("Error: ${e.message}")
                callback(Result.failure(FlutterError("channel-error", "Sign-in failed", e.message)))
            }
        }
    }

    override fun signOut(callback: (Result<Unit>) -> Unit) {
        checkInitialization()
        CoroutineScope(Dispatchers.Main).launch {
            try {
                when (val result = zendesk!!.logoutUser()) {
                    is ZendeskResult.Success -> {
                        callback(
                            Result.success(Unit)
                        )
                    }

                    is ZendeskResult.Failure -> {
                        callback(Result.failure(result.error))
                    }
                }
            } catch (e: Exception) {
                callback(
                    Result.failure(
                        FlutterError(
                            "channel-error", "Sign-out failed", e.message
                        )
                    )
                )
            }
        }
    }

    private fun checkInitialization() {
        if (zendesk == null) {
            throw FlutterError("channel-error", "Zendesk is not initialized", "")
        }
    }

    override fun getUnreadMessageCount(callback: (Result<Long>) -> Unit) {
        checkInitialization()
        CoroutineScope(Dispatchers.Main).launch {
            try {
                val unreadMessageCount = zendesk!!.messaging.getUnreadMessageCount()
                callback(Result.success(unreadMessageCount.toLong()))
            } catch (e: Exception) {
                callback(
                    Result.failure(
                        FlutterError(
                            "channel-error", "Failed to get unread message count", e.message
                        )
                    )
                )
            }
        }
    }

    override fun startListener() {
        checkInitialization()
        zendesk!!.addEventListener(zendeskEventListener)
    }

    override fun stopListener() {
        checkInitialization()
        zendesk!!.removeEventListener(zendeskEventListener)
    }

    override fun setLightColorRgba(colors: my.com.fromyourlover.pigeon.UserColors) {
        lightColors = UserColors(
            onPrimary = getAndroidColor(colors.onPrimary),
            onAction = getAndroidColor(colors.onAction),
            onMessage = getAndroidColor(colors.onMessage)
        )
    }

    override fun setDarkColorRgba(colors: my.com.fromyourlover.pigeon.UserColors) {
        darkColors = UserColors(
            onPrimary = getAndroidColor(colors.onPrimary),
            onAction = getAndroidColor(colors.onAction),
            onMessage = getAndroidColor(colors.onMessage)
        )
    }

    private fun getAndroidColor(color: my.com.fromyourlover.pigeon.RgbaColor?): Int? {

        if (color == null) return null
        return Color.valueOf(
            color.r.toFloat(), color.g.toFloat(), color.b.toFloat(), color.a.toFloat()
        ).componentCount
    }

    override fun enableLogging(enabled: Boolean) = Logger.setLoggable(enabled)

    override fun loggingEnabled(): Boolean = Logger.isLoggable()


}

File: ios/Classes/ZendeskPlusPlugin.swift
import Foundation
import Flutter
import UIKit
import ZendeskSDKMessaging
import ZendeskSDK
import ZendeskSDKLogger

public class ZendeskPlusPlugin: NSObject, FlutterPlugin, ZendeskHostApi {
    
    // MARK: - FlutterPlugin
    private var zendesk: Zendesk?
    var isloggedIn: Bool = false
    var factory = DefaultMessagingFactory()
    private static var zendeskListener: ZendeskListener?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = ZendeskPlusPlugin()
        ZendeskHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
        zendeskListener = ZendeskListener(binaryMessenger: registrar.messenger())
    }
    
    func initialized() throws -> Bool {
        return zendesk != nil
    }
    
    func signedIn() throws -> Bool {
        return isloggedIn
    }

    func initialize(androidChannelId: String?, iosChannelId: String?, completion: @escaping (Result<Void, Error>) -> Void) {
        // Ensure the iOS channel ID is provided
        guard let iosChannelId = iosChannelId else {
            let error = PigeonError(
                code: "invalid_ios_channel_id",
                message: "iOS channel ID is required",
                details: nil
            )
            completion(.failure(error))
            return
        }
        
        // Initialize Zendesk with the provided channel key
        Zendesk.initialize(withChannelKey: iosChannelId, messagingFactory: factory) {  result in
            DispatchQueue.main.async {
                switch result {
                case .success(let zendeskInstance):
                    // Store the initialized Zendesk instance
                    self.zendesk = zendeskInstance
                    print("Zendesk initialized successfully.")
                    completion(.success(()))
                case .failure(let error):
                    print("Zendesk initialization failed: \(error.localizedDescription)")
                    completion(.failure(error))
                }
            }
        }
    }
    
    func openChat() throws {
        // Ensure Zendesk is initialized
        guard let zendeskInstance = Zendesk.instance else {
            throw PigeonError(
                code: "zendesk_not_initialized",
                message: "Zendesk instance is not initialized",
                details: nil
            )
        }

        // Ensure messaging is available
        guard let messagingViewController = zendeskInstance.messaging?.messagingViewController() else {
            throw PigeonError(
                code: "messaging_not_available",
                message: "Zendesk messaging view controller is not available",
                details: nil
            )
        }

        // Get the root view controller
        guard let rootViewController = UIApplication.shared.keyWindow?.rootViewController else {
            throw PigeonError(
                code: "root_view_controller_not_found",
                message: "Root view controller is nil",
                details: nil
            )
        }

        // Wrap the messaging view controller in a navigation controller
        let navController = UINavigationController(rootViewController: messagingViewController)

        // Present the navigation controller
        DispatchQueue.main.async {
            if let presentedVC = rootViewController.presentedViewController {
                if presentedVC !== navController {
                    presentedVC.dismiss(animated: true) {
                        rootViewController.present(navController, animated: true, completion: nil)
                    }
                } else {
                    print("Zendesk messaging view controller is already presented")
                }
            } else {
                rootViewController.present(navController, animated: true, completion: nil)
            }
        }
    }
    
    func signIn(jwt: String, completion: @escaping (Result<ZendeskUser, Error>) -> Void) {
        guard let zendesk = Zendesk.instance else {
            completion(.failure(PigeonError(
                code: "zendesk_not_initialized",
                message: "Zendesk SDK is not initialized.",
                details: nil
            )))
            return
        }
        
        zendesk.loginUser(with: jwt) { result in
            DispatchQueue.main.async {
                switch result {
                case .success(let user):
                    let zendeskUser = ZendeskUser(id: user.id, externalId: user.externalId)
                    self.isloggedIn = true
                    completion(.success(zendeskUser))
                case .failure(let error):
                    print("Sign-in failed: \(error.localizedDescription)")
                    completion(.failure(error))
                }
            }
        }
    }

    
    func signOut(completion: @escaping (Result<Void, Error>) -> Void) {
        guard let zendesk = Zendesk.instance else {
            completion(.failure(PigeonError(
                code: "zendesk_not_initialized",
                message: "Zendesk instance is not initialized",
                details: nil
            )))
            return
        }

        zendesk.logoutUser { result in
            DispatchQueue.main.async {
                switch result {
                case .success:
                    self.isloggedIn = false
                    completion(.success(()))
                case .failure(let error):
                    // Propagate the error to the completion handler
                    completion(.failure(error))
                }
            }
        }
    }
    
    
    func getUnreadMessageCount(completion: @escaping (Result<Int64, Error>) -> Void) {
        // Retrieve unread message count
        let count = zendesk?.messaging?.getUnreadMessageCount() ?? 0
        completion(.success(Int64(count)))
    }

    func startListener() {
        
        zendesk?.addEventObserver(self) { event in
            switch event {
                case .unreadMessageCountChanged(let currentUnreadCount):
                    ZendeskPlusPlugin.zendeskListener?.onEvent(event: ZendeskEvent.unreadMessageCountChanged, completion: {_ in })

                    
                case .authenticationFailed(let error as NSError):
                    ZendeskPlusPlugin.zendeskListener?.onEvent(event: ZendeskEvent.authenticationFailed, completion: {_ in })
                
                    if error.code == 401 {
                        ZendeskPlusPlugin.zendeskListener?.onEvent(event: ZendeskEvent.jwtExpiredException, completion: {_ in })
                    }
                    
                case .conversationAdded(conversationId: let conversationId):
                    ZendeskPlusPlugin.zendeskListener?.onEvent(event: ZendeskEvent.conversationAdded , completion: {_ in })
                    
                case .connectionStatusChanged(connectionStatus: let connectionStatus):
                    ZendeskPlusPlugin.zendeskListener?.onEvent(event: ZendeskEvent.connectionStatusChanged , completion: {_ in })
                    
                case .sendMessageFailed(let error as NSError):
                    ZendeskPlusPlugin.zendeskListener?.onEvent(event: ZendeskEvent.sendMessageFailed , completion: {_ in })
                    
                @unknown default:
                    break
            }
        }
    }

    func stopListener() throws {
        // Stop listening for Zendesk events
        zendesk?.removeEventObserver(self)
    }

    func setLightColorRgba(colors: UserColors) {
        // Convert RgbaColor to UIColor
        let onPrimaryColor = UIColor.fromRgbaColor(colors.onPrimary ?? RgbaColor(r: 255.0, g: 0.0, b: 0.0, a: 1.0))
        let onMessageColor = UIColor.fromRgbaColor(colors.onMessage ?? RgbaColor(r: 255.0, g: 0.0, b: 0.0, a: 1.0))
        let onActionColor = UIColor.fromRgbaColor(colors.onAction ?? RgbaColor(r: 255.0, g: 0.0, b: 0.0, a: 1.0))

        // Create ZendeskSDKMessaging.UserColors with UIColor
        let userLightColors = ZendeskSDKMessaging.UserColors(
            onPrimary: onPrimaryColor,
            onMessage: onMessageColor,
            onAction: onActionColor
        )

        // Update the factory with the new colors
        self.factory = DefaultMessagingFactory(userLightColors: userLightColors)
    }

    func setDarkColorRgba(colors: UserColors) throws {
        // Convert RgbaColor to UIColor
        let onPrimaryColor = UIColor.fromRgbaColor(colors.onPrimary!)
        let onMessageColor = UIColor.fromRgbaColor(colors.onMessage!)
        let onActionColor = UIColor.fromRgbaColor(colors.onAction!)

        // Create ZendeskSDKMessaging.UserColors with UIColor
        let userColors = ZendeskSDKMessaging.UserColors(
            onPrimary: onPrimaryColor,
            onMessage: onMessageColor,
            onAction: onActionColor
        )

        // Update the factory with the new colors
        self.factory = DefaultMessagingFactory(userLightColors: ZendeskSDKMessaging.UserColors(onPrimary: nil, onMessage: nil, onAction: nil), userDarkColors: userColors)
    }

    func enableLogging(enabled: Bool) {
        Logger.enabled = enabled
    }

    func loggingEnabled() -> Bool {
        return true;
    }

}


extension UIColor {
    static func fromRgbaColor(_ rgbaColor: RgbaColor) -> UIColor {
        return UIColor(
            red: CGFloat(rgbaColor.r),
            green: CGFloat(rgbaColor.g),
            blue: CGFloat(rgbaColor.b),
            alpha: CGFloat(rgbaColor.a)
        )
    }
}
